Add stack graphs

jorendorff opened the pull request.

jorendorff committed:
hello world

jorendorff committed:
Use fewer nodes

jorendorff committed:
Lexical scopes.

jorendorff committed:
Non-working tests.

jorendorff committed:
Stop treating defs as uses.

jorendorff committed:
Type parameters; stop marking type defs as uses.

jorendorff committed:
identifiers no longer all need scopes

jorendorff committed:
Very basic support for fields.

jorendorff committed:
Make all tests pass.

jorendorff committed:
Implement shadowing.

jorendorff committed:
Don't crash on use-declarations.

jorendorff committed:
Support const generics except the ambiguous case.

The ambiguity is in GenericArgument context when the argument is a single
identifier.

jorendorff committed:
Fix captured patterns.

jorendorff committed:
Basic lifetime lookup support

jorendorff committed:
A few more basic lifetime tests.

jorendorff committed:
Fix and test for struct patterns.

jorendorff committed:
Fix and test for closures.

jorendorff committed:
Comments.

jorendorff committed:
Fix for tuple struct patterns.

jorendorff committed:
Add scripts I've been using. I expect to throw these away later.

jorendorff committed:
Resolve references to `self`.

jorendorff committed:
Trivial refactor of item rules.

jorendorff committed:
Support items in block scope.

jorendorff committed:
Bug fix and test for field names in struct expressions.

jorendorff committed:
Reorganize test directory.

Now it's part of a Rust project, so it's easy to test that the code is really
valid Rust code (which has tripped me up a few times): `cargo build --bins` in
the stack-graph-tests directory.

I was hoping emacs would turn on lsp-mode for these files, but it doesn't.

jorendorff committed:
Fix thinko

jorendorff committed:
Reimplement scope graph without parallel trees.

jorendorff committed:
Implement loop labels.

jorendorff committed:
Adjust test script command line to gitignore test output

hendrikvanantwerpen commented on queries/stack-graphs.tsg:
```diff
@@ -0,0 +1,818 @@
+;
+; Thus references to values can just push the identifier and go to the scope;
+; but references to types must push the type-identifier, then push `"%type"`,
+; then go to the scope.
```
I think this will work okay. I have always made all namespaces explicit (e.g., using `%val` and `%type`) so that all references and definitions have a similar structure in the graph. But I cannot immediately think of a scenario where one unguarded namespace would result in incorrect behavior.

hendrikvanantwerpen commented on queries/stack-graphs.tsg:
```diff
@@ -0,0 +1,818 @@
+;
+; Lifetimes have their own separate namespace in Rust, but we put them in the
+; type-namespace nodes. Since lifetimes always start with an apostrophe, there
+; can't be any confusion.
```
This should indeed be fine. I sometimes like to think about what the schema/types of a graph look like. Something like this would probably require a more expressive schema language to describe. But this is all speculative, so that consideration has no practical consequences.

hendrikvanantwerpen commented on queries/stack-graphs.tsg:
```diff
@@ -0,0 +1,818 @@
+
+;; Create the item scope.
+(block) @block {
+  ;; TODO: Figure out how to create these nodes conditionally. They're only
+  ;; needed if the block contains items; otherwise these can just be aliases.
+  ;; The reason it isn't done is I don't know how to use the query language
+  ;; to distinguish blocks that have items from those that don't.
```
You can distinguish empty and non empty blocks as follows. An empty block is matched using:
```
(block (_)*@children) {
  if (is-empty @children) {
    ; empty block
  } else {
    ; non-empty block
  }
}
```

If you prefer to have a separate rule for the non-empty block, you can write that without the `if` as follows:
```
(block (_)+) {
  ; non-empty block
}
```

jorendorff replied:
What I'd like to detect is if the block contains any Rust "items" (that is, declarations starting with `fn`, `struct`, `mod` etc., or imports) as opposed to just let-declarations, expression-statements, and/or a final expression. Very few blocks contain items.

This isn't important. But if the query language _can_ express this, knowing that would be very important!

hendrikvanantwerpen commented on queries/stack-graphs.tsg:
```diff
@@ -0,0 +1,818 @@
+}
+
+(block [
+  ;; This is every node type that can appear in a block as a statement, except
+  ;; `let_declaration`. Awkward but I don't know how else to write this.
```
This is indeed hard to express neatly in tree-sitter queries. You could probably do away with the explicit list if you are a bit more liberal with the amount of graph nodes that you create. You could set up nodes for each statement, and chain them, and link up the `.defs` of each statement to the correct node in the chain. Most of these will just not point to anything useful.

Our aim is that nodes and edges are cheap, so cheap that you don't feel the need to bend the rules to minimize their use. If they are not cheap enough, it's definitely something we'd try to solve generally in the library.

hendrikvanantwerpen commented on queries/stack-graphs.tsg:
```diff
@@ -0,0 +1,818 @@
+  edge t -> @ident.scope
+}
+
+;; The type-tag in a tuple_struct_pattern is a gruesome special case in rustc.
+;; It looks in both namespaces, preferring inner scopes to outer scopes, and
+;; when it finds both a value and a type in the same lexical scope, picking the
+;; type.
```
I tried to understand this one, but didn't manage yet. What I do for tricky cases that I cannot/do not support is write a test that shows the behavior and name it `.rs.skip`. This way it doesn't run, but it's still there as documentation.

hendrikvanantwerpen reviewed (commented):
Great work so far! I've looked through the TSG and the tests, and it all looks very sensible :+1: I love the comments, it made it quite easy to follow along. I've left some thoughts here and there, which may be useful to you.

I am assuming you set up the tests in this structure so you can also run them using Cargo?

jorendorff committed:
Cope with struct literals we can't trace yet.

jorendorff committed:
Slightly simplify block scope implementation.

jorendorff committed:
Comments and some very light refactoring.

jorendorff committed:
A comment.

jorendorff committed:
Support looking up types in `Iterator<Item=_>` context.

jorendorff committed:
Use less awkward ways of specifying parts of a match_pattern.

jorendorff committed:
Hook up `mod` items. Fix bug handling `let _ = ...;`.

The special pattern `_` parses as nothing. Probably many bugs.

jorendorff committed:
Trivial: sort some items

jorendorff committed:
First stab at implementing paths.

jorendorff committed:
Fix typo

jorendorff committed:
First test for paths.

jorendorff committed:
Hook up enum variants as declarations. Rough.

jorendorff committed:
Treat values like types, rather than hang them directly on scopes.

This should make paths to const and static items work, as well as paths in
the position marked by `(_)` in `(tuple_struct_pattern type: (_))` and
`(struct_pattern type: (_))`. Incomplete tests.

jorendorff committed:
rename a keyword

jorendorff committed:
Tests.

jorendorff committed:
Comment.

jorendorff committed:
Bind static items.

jorendorff committed:
Fix comment about loop labels.

jorendorff committed:
Straighten out parameters.

jorendorff committed:
Refactor pattern bindings.

jorendorff committed:
Refactor item bindings so they are @item.def, not @name.def.

jorendorff committed:
Recognize type context after `:` in let-declarations.

jorendorff committed:
Tag self, super, and crate in certain contexts as references.

jorendorff committed:
Define `self`, `super`, and `crate`.

jorendorff committed:
Basic imports.

jorendorff committed:
Names aren't expressions and don't need scopes.

jorendorff committed:
Structs with named fields are only in the type namespace.

jorendorff committed:
Add assertions to struct_pattern test

jorendorff committed:
Basic scope hookup for bracketed_type.

jorendorff committed:
`(constrained_type_parameter left: _)` is a binding context.

jorendorff committed:
`(constrained_type_parameter bounds: _)` needs a scope.

jorendorff committed:
`type_item` and `trait_item` are definitions.

jorendorff committed:
Hook up trait members and fully qualified method call syntax.

This is the barest minimum support for traits, and the rest will be a long time
coming! Fully qualified is the easy case.

jorendorff committed:
Passing tests for qualified access to enum variants.

jorendorff committed:
The right side of a type alias is a type-expression context.

jorendorff committed:
First multifile test.

There is plenty wrong with my approach here, but something works!

jorendorff committed:
Support cross-file `super`.

jorendorff committed:
Test file I found lying around.

jorendorff committed:
Support importing from other files via `mod foo;` items.

Basic cross-file imports work in more cases now. `mod xyzzy;` syntax is now
understood. Importing from `super::a::b` works. Arbitrarily nested directories
work.

jorendorff committed:
Change test script to expect the user to `cargo install` the tester.

I think this will be easier for anyone than copying a binary into a magic
location, and it allows me to use a non-released version of the tester via
`cargo install --features cli --path SRCDIR`.

jorendorff committed:
Fix comments explaining scopes and namespaces.

These have been wrong since I implemented `"%value"` nodes.

jorendorff committed:
Support resolving the last component of an import.

Introduces "%any".

jorendorff committed:
Use `;;` on initial comment (and some minor elaborations).

Emacs indentation seems to think `;` is for one-line comments, a nuisance.

jorendorff committed:
Big reorg collecting all the code about Rust paths in one place.

Lots and lots of juicy comments.

jorendorff committed:
Handle more `use` syntax: aliases and scoped use-lists.

Wildcards, leading `::`, `self` in imports, and unscoped use-lists are not
supported yet.

jorendorff committed:
Fix a few errors and warnings in tests.

jorendorff committed:
Add a unit test for an individual identifier in a scoped use-list.

jorendorff committed:
Passing test for self and super.

jorendorff committed:
Support `use { ... };` syntax (unqualified use-list).

jorendorff committed:
Support `use a::b::{self, c};` imports.

jorendorff committed:
Passing tests for `use crate::`.

jorendorff committed:
Support crate root imports.

jorendorff committed:
Passing test about generic function scopes.

jorendorff committed:
Don't crash on `use ::{anyhow};`.

jorendorff committed:
Passing tests for struct-like enum variants.

Importing them works!

jorendorff committed:
Merge branch 'master' into add-stack-graphs

jorendorff commented:
Heading upstream...
